fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,249 @@
+    /// The block hash is optional to allow for null blocks on Filecoin rootnet.
+    chain: TVar<im::OrdMap<BlockHeight, Option<V>>>,
+
+    /// Index votes received by height and hash, which makes it easy to look up
+    /// all the votes for a given block hash and also to verify that a validator
+    /// isn't equivocating by trying to vote for two different things at the
+    /// same height.
+    votes: TVar<im::OrdMap<BlockHeight, im::HashMap<V, im::HashSet<K>>>>,
+
+    /// Adding votes can be paused if we observe that looking for a quorum takes too long
+    /// and is often retried due to votes being added.
+    pause_votes: TVar<bool>,
+}
+
+impl<K, V> VoteTally<K, V>
+where
+    K: Clone + Hash + Eq + Sync + Send + 'static,
+    V: AsRef<[u8]> + Clone + Hash + Eq + Sync + Send + 'static,
+{
+    /// Initialize the vote tally from the current power table
+    /// and the last finalized block from the ledger.
+    pub fn new(power_table: Vec<(K, Weight)>, last_finalized_block: (BlockHeight, V)) -> Self {
+        let (height, hash) = last_finalized_block;
+        Self {
+            power_table: TVar::new(im::HashMap::from_iter(power_table)),
+            chain: TVar::new(im::OrdMap::from_iter([(height, Some(hash))])),
+            votes: TVar::default(),
+            pause_votes: TVar::new(false),
+        }
+    }
+
+    /// Check that a validator key is currently part of the power table.
+    pub fn known_validator(&self, validator_key: &K) -> Stm<bool> {
+        let pt = self.power_table.read()?;
+        // For consistency consider validators without power unknown.
+        match pt.get(validator_key) {
+            None => Ok(false),
+            Some(weight) => Ok(*weight > 0),
+        }
+    }
+
+    /// Calculate the minimum weight needed for a proposal to pass with the current membership.
+    pub fn quorum_threshold(&self) -> Stm<Weight> {
+        let total_weight: Weight = self.power_table.read().map(|pt| pt.values().sum())?;
+
+        Ok(total_weight * 2 / 3)
+    }
+
+    /// Return the height of the first entry in the chain.
+    pub fn last_finalized_height(&self) -> Stm<BlockHeight> {
+        self.chain
+            .read()
+            .map(|c| c.get_min().map(|(h, _)| *h).unwrap_or_default())
+    }
+
+    /// Add the next final block observed on the parent blockchain.
+    ///
+    /// Returns an error unless it's exactly the next expected height.
+    pub fn add_block(
+        &self,
+        block_height: BlockHeight,
+        block_hash: Option<V>,
+    ) -> StmResult<(), Error<K>> {
+        let mut chain = self.chain.read_clone()?;
+
+        // Check that we are extending the chain. We could also ignore existing heights.
+        if let Some((prev, _)) = chain.get_max() {
+            if block_height != prev + 1 {
+                return abort(Error::UnexpectedBlock(prev + 1, block_height));
+            }
+        }
+
+        chain.insert(block_height, block_hash);
+
+        self.chain.write(chain)?;
```

Comment:
Are you using immutable datastructures due to using the stm library? Copying them should be fast, but what is the performance if we need to do multiple times, like it seems we are doing here? Did you consider just using mutable datastructure combined with simple locks as an alternative?

Reply:
That's right, I use `im` because of the way the STM library works, by doing a `.clone()` when replacing values. Copying them is just makes another reference to the root, with modifications having logarithmic cost due to structural sharing. 

It's true that STM will not be as performant as a hand-built solution with mutable data structures and locking. It's a tradeoff for composability and the kind of API it supports, e.g. in https://github.com/consensus-shipyard/ipc/pull/590 I can combine queries that the component wasn't completely aware of, or in https://github.com/consensus-shipyard/ipc/pull/584 I can query two STM components and know they are consistent with each other, the updates happen in lockstep with the syncers own data structures. 

A non-locking alternative of using channels with mutable data structures would surely be faster, but wouldn't compose and could be inconsistent across these isolated islands of state.

Reply:
STM is used prominently in [Cardano](https://github.com/replica-io/replica-io/issues/26#issuecomment-1806433718). At the time I was interested in what we could take on from that project and pursued it in Rust. I like using it, but maybe in Rust it's not as nice because for example of the need to clone the inputs. Maybe it's a bit of an overkill for us, but I recommended to Will we try it when he was doing traditional locking and I found some race conditions IIRC. I think it's less tedious then the pattern with channels and clients and response channels in events, like inside the IPLD resolver service and client, and it makes me think a bit. 

I figured in this use case where we have to vote on blocks that are produced every 30 seconds it should be fine. But if the parent would be Fendermint too with 1s block time, gossiping votes could be overwhelming and we should possibly reduce the voting frequency to every N-th block.